### PR TITLE
Edit clap solver to use dynamic lsap code

### LINKED ISSUE #26: Modify `costs` function to use our dynamic stuff.

### PR DESCRIPTION
**Pull request by qinyichen**
Resolves #26

### CODE REVIEW

File: tests/test_lsap.py [hunk 1]
```diff
@@ -94,6 +94,13 @@ def test_linear_sum_assignment_input_validation():
     eye[:, 0] = np.inf
     assert_raises(ValueError, linear_sum_assignment, eye)
 
+    # Note that solve_lsap requires that num_rows cannot exceed num_cols
+    cost_matrix = np.array([[0     , np.Inf, np.Inf],
+                            [np.Inf,      0, np.Inf],
+                            [np.Inf, np.Inf,      0],
+                            [0     ,      0,      0]])
+    assert_raises(ValueError, solve_lsap, cost_matrix)
```
Comment (jdmoorman) [inline tests/test_lsap.py:102]: Why should this raise a value error?

File: tests/test_lsap.py [hunk 2]
```diff
@@ -94,6 +94,13 @@ def test_linear_sum_assignment_input_validation():
     eye[:, 0] = np.inf
     assert_raises(ValueError, linear_sum_assignment, eye)
 
+    # Note that solve_lsap requires that num_rows cannot exceed num_cols
```
Comment (jdmoorman) [inline tests/test_lsap.py:97]: We should check the shape of the matrix at the start and operate on the transpose if it is the wrong shape.

File: src/laptools/clap.py [hunk 3]
```diff
@@ -96,41 +104,54 @@ def costs(cost_matrix):
...
-            # This calculation will end up being wrong for the columns in
-            # lsap_col_idxs[sub_col_ind]. This is because the constraint in
-            # row i in these columns will conflict with the sub assignment.
-            # These miscalculations are corrected later.
-            total_costs[i, :] = cost_matrix[i, :] + sub_total_cost
+
+        # Solve the lsap with row i omitted. For the majority of
+        # constraints on row i's assignment, this will not conflict with
+        # the constraint. When it does conflict, we fix the issue later.
+        sub_ind = ~one_hot(i, n_rows)
```
Comment (jdmoorman) [inline src/laptools/clap.py:120]: Add
```python
sub_cost_matrix = cost_matrix[sub_ind, :]
```
here to use later.

File: src/laptools/clap.py [hunk 4]
```diff
@@ -146,16 +167,33 @@ def costs(cost_matrix):
...
+                sub_lsap_row4col = new_lsap_row4col.copy()
+                # Note that row i is assigned to some column. Need to remove
+                # that from the assignment.
+                new_freed_j = new_lsap_col4row[i]
+                sub_lsap_row4col[new_freed_j] = -1
+                sub_lsap_row4col[sub_lsap_row4col > i] -= 1
+
+                # Otherwise, solve the lsap with stolen_j removed
+                _, col4row_with_removed_col, _, _ = solve_lsap_with_removed_col(
+                    cost_matrix[sub_ind, :],
```
Comment (jdmoorman) [inline src/laptools/clap.py:179]: Use `sub_cost_matrix`

File: src/laptools/clap.py [hunk 5]
```diff
@@ -146,16 +167,33 @@ def costs(cost_matrix):
...
-                sub_cost_matrix = cost_matrix[:, potential_cols]
-                sub_j = np.argwhere(potential_cols == stolen_j)[0]
-                total_cost = cost(i, sub_j, sub_cost_matrix)
-                total_costs[i, stolen_j] = total_cost
+                sub_lsap_row4col = new_lsap_row4col.copy()
+                # Note that row i is assigned to some column. Need to remove
+                # that from the assignment.
+                new_freed_j = new_lsap_col4row[i]
+                sub_lsap_row4col[new_freed_j] = -1
+                sub_lsap_row4col[sub_lsap_row4col > i] -= 1
```
Comment (jdmoorman) [inline src/laptools/clap.py:175]: Is any of this things that could be done outside the loop?

File: src/laptools/clap.py [hunk 6]
```diff
@@ -73,20 +78,23 @@ def costs(cost_matrix):
     # Find the best lsap assignment from rows to columns without constrains.
     # Since there are at least as many columns as rows, row_idxs should
     # be identical to np.arange(n_rows). We depend on this.
-    row_idxs, lsap_col_idxs = lap(cost_matrix)
+    row_idxs = np.arange(n_rows)
+    lsap_row4col, lsap_col4row, u, v = solve_lsap(cost_matrix)
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:82]: Yes at line 126.

File: src/laptools/clap.py [hunk 7]
```diff
@@ -73,20 +78,23 @@ def costs(cost_matrix):
...
-    used = np.isin(sorted_costs_ind[:, : n_rows + 1], lsap_col_idxs)
-    first_unused = sorted_costs_ind[row_idxs, np.argmax(~used, axis=1)]
-    potential_cols = np.union1d(lsap_col_idxs, first_unused)
+    # TODO: commented out potential_cols for now. Bring back later.
+
+    # # When a row has its column stolen by a constraint, these are the columns
+    # # that might come into play when we are forced to resolve the assignment.
+    # used = np.isin(sorted_costs_ind[:, : n_rows + 1], lsap_col4row)
+    # first_unused = sorted_costs_ind[row_idxs, np.argmax(~used, axis=1)]
+    # potential_cols = np.union1d(lsap_col4row, first_unused)
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:97]: If we use potential_cols we can deal with less cols in solve_lsap_with_removed_col. I'm not sure how complicated the indexing will be though.

File: src/laptools/clap.py [hunk 8]
```diff
@@ -146,16 +167,33 @@ def costs(cost_matrix):
...
+                    sub_lsap_row4col,
+                    new_lsap_col4row[sub_ind],
+                    new_u[sub_ind],  # dual variable associated with rows
+                    new_v.copy(),  # dual variable associated with cols
+                )
+
+                total_costs[i, stolen_j] = (
+                    cost_matrix[i, stolen_j]
+                    + cost_matrix[sub_ind, col4row_with_removed_col].sum()
+                )
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:190]: black does that

File: tests/test_lsap.py (hunk 1 above)
Comment (qinyichen) [reply tests/test_lsap.py:102]: solve_lsap need num_rows <= num_cols.

File: tests/test_lsap.py (hunk 2 above)
Comment (qinyichen) [reply tests/test_lsap.py:97]: linear_sum_assignment does that. solve_lsap does not. (They are two different functions in dynamic_lsap.py)

File: src/laptools/clap.py [hunk 9]
```diff
@@ -96,41 +104,54 @@ def costs(cost_matrix):
...
+            cost_matrix,
+            i,
+            lsap_row4col.copy(),
+            lsap_col4row.copy(),
+            u.copy(),
+            v.copy(),
+        )
+
+        sub_total_cost = cost_matrix[sub_ind, new_lsap_col4row[sub_ind]].sum()
+        col_idxs[sub_ind] = new_lsap_col4row[sub_ind]
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:133]: Fixed. Also about to change a bunch of variable names.

File: src/laptools/clap.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:97]: Deleting this for now but we can add them back later.

File: src/laptools/clap.py [hunk 10]
```diff
@@ -96,41 +104,54 @@ def costs(cost_matrix):
...
-            # row i in these columns will conflict with the sub assignment.
-            # These miscalculations are corrected later.
-            total_costs[i, :] = cost_matrix[i, :] + sub_total_cost
+
+        # Solve the lsap with row i omitted. For the majority of
+        # constraints on row i's assignment, this will not conflict with
+        # the constraint. When it does conflict, we fix the issue later.
+        sub_ind = ~one_hot(i, n_rows)
+
+        # TODO: Is there a better way than passing in copies here?
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:122]: Added a flag that deals with this.

File: src/laptools/clap.py [hunk 11]
```diff
@@ -146,16 +167,33 @@ def costs(cost_matrix):
...
+                sub_lsap_row4col[sub_lsap_row4col > i] -= 1
+
+                # Otherwise, solve the lsap with stolen_j removed
+                _, col4row_with_removed_col, _, _ = solve_lsap_with_removed_col(
+                    cost_matrix[sub_ind, :],
+                    stolen_j,
+                    sub_lsap_row4col,
+                    new_lsap_col4row[sub_ind],
+                    new_u[sub_ind],  # dual variable associated with rows
+                    new_v.copy(),  # dual variable associated with cols
```
[the remark this answers is not in the document]
Comment (qinyichen) [reply src/laptools/clap.py:184]: No longer applicable.

File: src/laptools/clap.py (hunk 5 above)
Comment (qinyichen) [reply src/laptools/clap.py:175]: Noted. Will fix.

**Pull request merged**